libraries/ESP8266WiFiMesh/src/MeshBackendBase.cpp

```diff
@@ -0,0 +1,294 @@
+/*
+  MeshBackendBase
+ 
+  Copyright (c) 2015 Julian Fell and 2019 Anders Löfgren. All rights reserved.
+ 
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include "MeshBackendBase.h"
+#include "TypeConversionFunctions.h"
+#include "MutexTracker.h"
+
+#include <assert.h>
+
+MeshBackendBase *MeshBackendBase::apController = nullptr;
+
+bool MeshBackendBase::_scanMutex = false;
+
+bool MeshBackendBase::_printWarnings = true;
+
+MeshBackendBase::MeshBackendBase(requestHandlerType requestHandler, responseHandlerType responseHandler, networkFilterType networkFilter, mesh_backend_t classType)
+{
+  setRequestHandler(requestHandler);
+  setResponseHandler(responseHandler);
+  setNetworkFilter(networkFilter);
+  setClassType(classType);
+}
+
+MeshBackendBase::~MeshBackendBase()
+{
+  deactivateAP();
+}
+
+void MeshBackendBase::setClassType(mesh_backend_t classType)
+{
+  _classType = classType;
+}
+
+mesh_backend_t MeshBackendBase::getClassType() {return _classType;}
+
+void MeshBackendBase::activateAP()
+{
+  // Deactivate active AP to avoid two servers using the same port, which can lead to crashes.
+  if(MeshBackendBase *currentAPController = MeshBackendBase::getAPController())
+    currentAPController->deactivateAP();
+
+  activateAPHook();
+  
+  WiFi.mode(WIFI_AP_STA); 
+
+  apController = this;
+}
+
+void MeshBackendBase::activateAPHook()
+{
+  WiFi.softAP( getSSID().c_str(), getMeshPassword().c_str(), getWiFiChannel(), getAPHidden() ); // Note that a maximum of 8 TCP/IP stations can be connected at a time to each AP, max 4 by default.
+}
+
+void MeshBackendBase::deactivateAP()
+{
+  if(isAPController())
+  {
+    deactivateAPHook();
+
+    WiFi.softAPdisconnect();
+    WiFi.mode(WIFI_STA);
+
+    // Since there is no active AP controller now, make the apController variable point to nothing. 
+    apController = nullptr;
+  }
+}
+
+void MeshBackendBase::deactivateAPHook()
+{
+}
+
+void MeshBackendBase::restartAP()
+{
+  deactivateAP();
+  yield();
+  activateAP();
+  yield();
+}
+
+MeshBackendBase *MeshBackendBase::getAPController()
+{
+  return apController; 
+}
+
+bool MeshBackendBase::isAPController()
+{
+  return (this == getAPController());
+}
+
+void MeshBackendBase::setWiFiChannel(uint8 newWiFiChannel)
+{
+  assert(1 <= newWiFiChannel && newWiFiChannel <= 13);
```

Comment:
Done, but the country API seems possibly broken for it never changes the allowed channels. Anyhow, the default values returned are the same as the previous ones used (1 and 13), so I'm keeping the API included in case it is fixed in the future.